persist/fs/commitlog/reader.go

```diff
@@ -0,0 +1,279 @@
+	index := int(r.info.Index)
+	return start, duration, index, nil
+}
+
+func (r *reader) Read() (
+	series m3db.CommitLogSeries,
+	datapoint m3db.Datapoint,
+	unit xtime.Unit,
+	annotation []byte,
+	resultErr error,
+) {
+	if err := r.read(&r.log); err != nil {
+		resultErr = err
+		return
+	}
+
+	if len(r.log.Metadata) != 0 {
+		if err := proto.Unmarshal(r.log.Metadata, &r.metadata); err != nil {
+			resultErr = err
+			return
+		}
+		r.metadataLookup[r.log.Idx] = m3db.CommitLogSeries{
+			UniqueIndex: r.log.Idx,
+			ID:          r.metadata.Id,
+			Shard:       r.metadata.Shard,
+		}
+	}
+
+	metadata, ok := r.metadataLookup[r.log.Idx]
+	if !ok {
+		resultErr = errCommitLogReaderMissingLogMetadata
+		return
+	}
+
+	series = metadata
+	datapoint = m3db.Datapoint{
+		Timestamp: time.Unix(0, r.log.Timestamp),
+		Value:     r.log.Value,
+	}
+	unit = xtime.Unit(byte(r.log.Unit))
+	annotation = r.log.Annotation
+	return
+}
+
+func (r *reader) read(message proto.Message) error {
+	// Read size of message
+	size, err := binary.ReadUvarint(r.chunkReaderRef)
+	if err != nil {
+		return err
+	}
+
+	// Extend buffer as necessary
+	if len(r.dataBuffer) < int(size) {
+		diff := int(size) - len(r.dataBuffer)
+		for i := 0; i < diff; i++ {
+			r.dataBuffer = append(r.dataBuffer, 0)
+		}
+	}
+
+	// Size the target buffer for reading and unmarshalling
+	buffer := r.dataBuffer[:size]
+
+	// Read message
+	if _, err := r.chunkReaderRef.Read(buffer); err != nil {
+		return err
+	}
+	if err := proto.Unmarshal(buffer, message); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *reader) Close() error {
+	if r.chunkReader.fd == nil {
+		return nil
+	}
+
+	if err := r.chunkReader.fd.Close(); err != nil {
+		return err
+	}
+
+	r.chunkReader.fd = nil
+	r.metadataLookup = make(map[uint64]m3db.CommitLogSeries)
+	return nil
+}
+
+type chunkReader struct {
+	fd        *os.File
+	buffer    *bufio.Reader
+	remaining int
+	charBuff  []byte
+}
+
+func (r *chunkReader) Read(p []byte) (int, error) {
+	size := len(p)
+	read := 0
+	// Check if requesting for size larger than this chunk
+	if r.remaining < size {
+		// Copy any remaining
+		if r.remaining > 0 {
+			n, err := r.buffer.Read(p[:r.remaining])
+			r.remaining -= n
+			read += n
+			if err != nil {
+				return read, err
+			}
+		}
+
+		// Read next header
+		if err := r.readHeader(); err != nil {
+			return read, err
+		}
+
+		// Reset read target
+		p = p[read:]
+		size = len(p)
+
+		// Perform consecutive read(s)
+		n, err := r.Read(p)
+		read += n
+		return read, err
+	}
+
+	n, err := r.buffer.Read(p)
+	r.remaining -= n
+	read += n
+	return read, err
+}
+
+func (r *chunkReader) ReadByte() (c byte, err error) {
+	if _, err := r.Read(r.charBuff); err != nil {
+		return byte(0), err
+	}
+	return r.charBuff[0], nil
+}
+
+func (r *chunkReader) reset(fd *os.File) {
+	r.fd = fd
+	r.buffer.Reset(fd)
+	r.remaining = 0
+	r.charBuff = r.charBuff[:0]
+	r.charBuff = append(r.charBuff, 0)
```

Comment:
Well there is no `newChunkReader(...)` method so I was accounting for the case where its uninitialized.  I'm fixing this by just creating a new `newChunkReader` for initializing the `chunkReader`.